lib/backend/pgbk/pgbk.go

```diff
@@ -0,0 +1,580 @@
+					return backend.Item{}, err
+				}
+				return backend.Item{
+					Key:     key,
+					Value:   value,
+					Expires: time.Time(expires).UTC(),
+					// Revision: crockfordLowercaseBase32.EncodeToString(revision.Bytes[:]),
+				}, nil
+			})
+			return trace.Wrap(err)
+		})
+
+		if err := b.pool.SendBatch(ctx, batch).Close(); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return items, nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &backend.GetResult{Items: items}, nil
+}
+
+// Delete implements [backend.Backend].
+func (b *Backend) Delete(ctx context.Context, key []byte) error {
+	deleted, err := pgcommon.Retry(ctx, b.log, func() (bool, error) {
+		tag, err := b.pool.Exec(ctx,
+			"DELETE FROM kv WHERE key = $1 AND (expires IS NULL OR expires > now())", key)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		return tag.RowsAffected() > 0, nil
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !deleted {
+		return trace.NotFound("key %q does not exist", key)
+	}
+	return nil
+}
+
+// DeleteRange implements [backend.Backend].
+func (b *Backend) DeleteRange(ctx context.Context, startKey []byte, endKey []byte) error {
+	// this is the only backend operation that might affect a disproportionate
+	// amount of rows at the same time; in actual operation, DeleteRange hardly
+	// ever deletes more than dozens of items at once, and logical decoding
+	// starts having performance issues when a transaction affects _thousands_
+	// of rows at once, so we're good here (but see [Backend.backgroundExpiry])
+	if _, err := pgcommon.Retry(ctx, b.log, func() (struct{}, error) {
+		_, err := b.pool.Exec(ctx,
+			"DELETE FROM kv WHERE key BETWEEN $1 AND $2",
+			startKey, endKey,
+		)
+		return struct{}{}, trace.Wrap(err)
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// KeepAlive implements [backend.Backend].
+func (b *Backend) KeepAlive(ctx context.Context, lease backend.Lease, expires time.Time) error {
+	revision := newRevision()
+	updated, err := pgcommon.Retry(ctx, b.log, func() (bool, error) {
+		tag, err := b.pool.Exec(ctx,
+			"UPDATE kv SET expires = $1, revision = $2 WHERE key = $3 AND (expires IS NULL OR expires > now())",
+			zeronull.Timestamptz(expires.UTC()), revision, lease.Key)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		return tag.RowsAffected() > 0, nil
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !updated {
+		return trace.NotFound("key %q does not exist", lease.Key)
+	}
+	return nil
+}
+
+// NewWatcher implements [backend.Backend].
+func (b *Backend) NewWatcher(ctx context.Context, watch backend.Watch) (backend.Watcher, error) {
+	return b.buf.NewWatcher(ctx, watch)
+}
+
+// CloseWatchers implements [backend.Backend].
+func (b *Backend) CloseWatchers() { b.buf.Clear() }
+
+// Clock implements implements [backend.Backend].
+func (b *Backend) Clock() clockwork.Clock { return clockwork.NewRealClock() }
```

Comment:
Can you comment on why this is always the real clock? I assume that it's because the backend is also using the DB time, so mocking would not work.

Reply:
641c39cfca

Reply:
in a previous version we shipped the local fake clock time to the database on every transaction, but using the server time has the significant advantage of getting all auth servers to agree about which items exist and which are expired, in spite of any possible clock drift